pkg/plugins/diff/service.go

```diff
@@ -0,0 +1,89 @@
+package tablediff
+
+import (
+	"context"
+	"time"
+
+	"github.com/treeverse/lakefs/pkg/plugins/internal"
+)
+
+type DiffEntry struct {
+	Version          string
+	Timestamp        time.Time
+	Operation        string
+	OperationContent map[string]string
+}
+
+type Response struct {
+	Diffs []DiffEntry
+}
+
+type RefPath struct {
+	Ref  string
+	Path string
+}
+
+type TablePaths struct {
+	LeftTablePath  RefPath
+	RightTablePath RefPath
+	BaseTablePath  RefPath
+}
+
+type S3Creds struct {
+	Key      string
+	Secret   string
+	Endpoint string
+}
+
+type Params struct {
+	TablePaths TablePaths
+	S3Creds    S3Creds
+}
+
+type Differ interface {
+	Diff(context.Context, Params) (Response, error)
+}
+
+// Service is responsible for registering new Differ plugins and executing them at will.
+// After initializing a Service, the CloseClients method should be called at some point to close gracefully all
+// remaining plugins.
+type Service struct {
+	pluginHandler  internal.Handler[Differ, internal.HCPluginProperties]
+	closeFunctions []func()
+}
+
+// NewService is used to initialize a new Differ service. The returned function is a closing function for the service.
+func NewService() (*Service, func()) {
+	service := &Service{
+		pluginHandler:  internal.NewManager[Differ](),
+		closeFunctions: make([]func(), 0),
+	}
+	return service, service.Close
+}
+
+func (s *Service) RunDiff(ctx context.Context, diffType string, diffParams Params) (Response, error) {
+	d, closeClient, err := s.pluginHandler.LoadPluginClient(diffType)
+	if err != nil {
+		return Response{}, err
+	}
+	if closeClient != nil {
+		s.closeFunctions = append(s.closeFunctions, closeClient)
```

Comment:
This seems thread-unsafe. Also, for every diff type do we have multiple clients? Because we _will_ add multiple close functions for every diff type.